server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -983,8 +994,47 @@ private Engine.IndexResult applyIndexOperation(
             verifyNotClosed(e);
             return new Engine.IndexResult(e, version, opPrimaryTerm, seqNo, sourceToParse.id());
         }
+        Engine.IndexResult indexResult = index(engine, operation);
+        if (indexResult.getResultType().equals(Engine.Result.Type.SUCCESS) && origin.equals(Engine.Operation.Origin.PRIMARY)) {
+            updateFieldHasValueWithFieldsFromParsedDoc(operation.parsedDoc());
+        }
+        return indexResult;
+    }
+
+    private void updateFieldHasValueWithFieldsFromParsedDoc(ParsedDocument document) {
+        for (LuceneDocument doc : document.docs()) {
+            for (IndexableField field : doc.getFields()) {
+                setFieldHasValue(field.name());
```

Comment:
Thanks for the input @jpountz, looking into it!

Reply:
> My suggestion is about treating them as one, if we find that _feature has value then we add all the rank_feature even if some could be empty.

That sounds sensible.

It should be possible to check which fields exist by reading the terms dictionary of the `_feature` field, where each term is a field value. But then we're requiring touching the disk to check which fields exist (unlike field infos which are in heap), which I'm not too comfortable with.

Reply:
> I have the change ready and it uses afterRefresh mechanics.

I'm not too familiar with `afterRefresh`, is it guaranteed to run before the new point-in-time view of the shard is published, so that searches could not see new fields before `_field_caps`?

Reply:
Answering my own question: `afterRefresh` runs after the new point-in-time view is published, so `_field_caps` may have some lag. Should we use `beforeRefresh` instead and fix it to take the new point-in-time view as an argument?

Reply:
Thanks @jpountz, I'll probably have find a better place then. From the tests it doesn't seem that the lag is major, but still a "bug". Moving this logic inside afterRefresh meant also that we make _filed_caps with non-empty fields only available to the users when such fields are available for search, that is perfect for the main user for this feature that would be kibana.

Reply:
I don't see a problem here, if I understand correctly, `index_filter` worked like this already. And recently we introduced caching for data view field_caps requests, where the data view cache is evaluated in the background, optimizing for the speed of data delivery, with the acceptable trade off, that the fields may not be up to date every time. So if there's a temporary inconsistency between fields in field_caps or search requests, this shouldn't be a problem in Kibana.

Reply:
Thanks @kertal